### PR TITLE
metaprogramming

### PR DESCRIPTION
**Pull request by DaanVanHauwermeiren**
metaprogramming !

This notebook contains:

1. program representation
2. symbols
3. quoting
4. splatting
5. evaluating expressions
6. basics of macros
7. why macros?
8. building macros
9. obligatory thing about  glorious multiple dispatch
10. code generation (+ example from JuMP)
11. non-standard string literals (+ example from Markdown and BioSequences)
12. some common macros

### CODE REVIEW

**Comment by DaanVanHauwermeiren** [comment]
Note: force push because of bad credentials

**Comment by MichielStock** [comment]
Nice. A few remarks
- make sure that the main features of metaprogramming are clear for the less technically inclined:
    - modifying code before it runs
    - adding new features/syntax
    - processing strings at compile time
- notebook has a lot of errors, at least partly due to macros are difficult to construct the dependency tree for. Some examples I can't get to work
- your `sayhello()`: maybe compute/print something before the expression is generated (only done at compile)
- you can define `sayhello()` in separate cells, strange...
- don't make it too long ;-)

**Comment by DaanVanHauwermeiren** [comment]
forgot to add new dependencies to `Project.toml`, see 83b99fb. All examples should work with these dependencies in place.

**Comment by DaanVanHauwermeiren** [comment]
I choose to remove the part on macro hygiene in 7aac5b8. Not because it is not interesting, but because it is very technical. I would rather spend some more time on examples than on technical peculiarities.

**Comment by DaanVanHauwermeiren** [comment]
I think, this is more or less finished now. If you have some suggestions, I'd love to hear them

File: notebooks/day2/metaprogramming.jl [hunk 1]
```diff
@@ -0,0 +1,1103 @@
...
+# ╔═╡ c42d2d6a-5e67-11eb-1564-2dcb0e91fb19
+@assert 1 == 0
+
+# ╔═╡ 376fe8e4-5e68-11eb-1d7d-c9b1945b133f
+md"""
+Now `@assert` has two modes of operation, depending upon the number of arguments it receives!
+If there's only one argument, the tuple of expressions captured by `msgs` will be empty and it
+will behave the same as the simpler definition above. But now if the user specifies a second argument,
+it is printed in the message body instead of the failing expression. You can inspect the result
+of a macro expansion with the aptly named `@macroexpand`
```
Comment (Beramos) [inline notebooks/day2/metaprogramming.jl:440]: There is something strange with this sentence.

**Review by Beramos — APPROVED** [review]

File: notebooks/day2/metaprogramming.jl (hunk 1 above)
Comment (Beramos) [reply notebooks/day2/metaprogramming.jl:445]: @DaanVanHauwermeiren this PR can be merged after fixing this, I'm too stupid to construct a proper sentence from this

Comment (DaanVanHauwermeiren) [reply notebooks/day2/metaprogramming.jl]: solved in f85d321

**Pull request merged**